### ISSUE TITLE
Question to EEGNet models

### ISSUE THREAD

**Issue by NeoMW** [braindecode/braindecode#186]
Since I'm really new to EEG and Deep learning, I'm experimenting with your example codes.
Especially with your notebook "plot_bcic_iv_2a_moabb_trial".
I wanted to try out your EEGNet models (the ShallowFBCSPNet,Deep4Net worked just fine), so I naively changed the model in the notebook to EEGNetv1 with the BNCI2014001 dataset.

Everything up to fitting the classifier worked just fine.
I get the following error message:
"Expected 4-dimensional input for 4-dimensional weight [16, 22, 1, 1], but got 3-dimensional input of size [64, 22, 1125] instead"
I'm really lost on what I'm doing wrong, could you give me a hint why this error is occurring or what I need to take into account? 

Thank's you for your Answer!

**Comment by hubertjb** [comment]
Hi @NeoMW, it looks like the network expects the input to have 4 dimensions, but your input only has 3. Quickly looking at the code, it looks like ShallowFBCSPNet and Deep4Net both have a `self.add_module("ensuredims", Ensure4d())` line that EEGNet doesn't have, and which should take care of adding a fourth dimension to the input automatically. @robintibor does that make sense?

**Comment by robintibor** [comment]
there is no good reason for this, let's fix it!

**Comment by robintibor** [comment]
So started a fix in https://github.com/braindecode/braindecode/pull/187 . will merge tomorrow

**Comment by NeoMW** [comment]
Hi, @robintibor @hubertjb many thanks for your helpful and fast response. I'm excited about the patch.

**Comment by NeoMW** [comment]
Hi, @robintibor @hubertjb I made the "ensure dimension change" locally, and noticed that the EEGNets throw "Calculated padded input size per channel: (8 x 942). Kernel size: (13 x 929). Kernel size can't be greater than actual input size" for both nets when I call the  "get_output_shape" function if I set "final_conv_length" to "auto". Is there any quick fix for this?  ^^

**Comment by robintibor** [comment]
So I have updated PR further, should be ready to merge hopefully.

Regarding your question, this indicates the n_in_times is too small, e.g., try this it should work:

``python
from braindecode.models.util import get_output_shape

n_channels = 2
n_in_times = 200
model = EEGNetv1(n_channels, n_classes, input_window_samples=n_in_times)
get_output_shape(model, n_channels, n_in_times)
``

Do you have any reason to prefer EEGNetv1 over EEGNetv4? In most cases v4 should be similar or better I think.

**Comment by NeoMW** [comment]
No, I don't have a reason, in that case, I will use EEGNetV4.
n_in_times is already at 1000, should I put that higher? 
For EEGNetv4, if I input small values for final_conv_length  it somehow works. 
Many thanks for your help.^^

**Comment by bokey007** [comment]
Hi, 

I was trying to understand how brain-decode works.

I think the training dataset created from windows_dataset returns a tuple of length 3

and the first element in this tuple represents data to be passed to model (and has shape of (channels, time_points)) 

Here i am not able to understand, why(/how) 4d input is passed?

since i was hoping for input of form (batch_size, channels, time_points)  

Looking forward to your help.

Thanks and regards 
bokey

**Comment by robintibor** [comment]
Hil,

in general in Pytorch, datasets you get individual items like in our case channels x timepoints of one window. To get batches you use dataloaders that create batches from the datasets.

https://pytorch.org/tutorials/beginner/basics/data_tutorial.html explains it fairly well I think

**Comment by bokey007** [comment]
ohh. seems like pytorch expects a volume as input, thus  -- > input had to be in the form of (batch_size, 1, channels, temp_point) 
--> and thus 4d input

@robintibor --> thanks for the response.

but i still don't understand how/which object/class implements pytorch dataset/dataloder in brain-decode. Is it windows_dataset or EEGClassifier ?

**Comment by robintibor** [comment]
PyTorch does not expect any specific input, it depends on the model what input it expects :)

Regarding datasets/dataloaders, you have to distinguish between datasets and dataloaders. 
In Braindecode, we provide a lot of dataset functionality like WindowsDatasets etc.. These Braindecode-Pytorch Datasets can be indexed and return one element from the dataset (e.g., a window of EEG data).

Dataloaders you just use regular PyTorch Dataloader, we have not implemented any specific Dataloader to use here, although we now prevent some samplers (which can be used together with the PyTorch Dataloder to determine which elements come in one batch) https://github.com/braindecode/braindecode/tree/master/braindecode/samplers

So, in summary, Braindecode gives you PyTorch datasets that you can use together with regular PyTorch dataloaders.

EEGClassifier internally creates the dataloader already, that's why you don't have to construct a dataloader explicitly if you use EEGClassifier.
